### PR TITLE
Option to configure proxy cache TTL

### PR DESCRIPTION
**Pull request by manishtomar**
Currently when registry is run as proxy it tries to cleanup unused blobs from its cache after 7 days which is hard-coded. This PR makes that value configurable. One can disable caching using this option. This allows one to temporarily mitigate
https://github.com/docker/distribution/issues/2367 bug.

### CODE REVIEW

**Comment by manishtomar** [comment]
With recent TOS change in Docker a lot more people may run this project as a mirror and that bug is a blocker in using that.

File: docs/configuration.md
Comment (thaJeztah) [inline docs/configuration.md:1121]: Should we keep the default same as the current value? (7 days)? I that case, possibly use some special value for disabling

File: registry/proxy/proxymanifeststore.go
Comment (thaJeztah) [inline registry/proxy/proxymanifeststore.go:14]: I see the original todo mentioned respecting cache-control headers; that still something we want to add at some point?

**Comment by thaJeztah** [comment]
@dmcgowan ptal

**Comment by billchenchina** [comment]
Any updates on this?
/cc @dmcgowan

File: registry/proxy/proxymanifeststore.go
Comment (milosgajdos) [reply registry/proxy/proxymanifeststore.go:14]: The `TODO` comment actually says:  "...instead of hardcoding an arbitrary chosen constant here, we could *either* pick it from cache headers or we can make move it to config". Given this PR moves the TTL value to config, I think this comment no longer applies.

File: docs/configuration.md
Comment (milosgajdos) [reply docs/configuration.md:1121]: In this PR implementation, if no config is specified there is no expiration. I like this approach as it forces explicit intent by the operator, rather than hiding some opaque behaviour. As long as the behaviour is documented and we no longer pick arbitrary values I'm happy with this.

**Review by milosgajdos — APPROVED** [review]
LGTM

File: docs/configuration.md
Comment (manishtomar) [reply docs/configuration.md:1121]: One thing to note is that operators who might be used to seeing reduced space will see it growing unbounded after this PR is deployed. Considering most of the operators would've encountered this bug after 7 days I think its fine to have this backward incompatible change.

Comment (wy65701436) [reply docs/configuration.md:1121]: If no default value set here, does it impact the upgrade scenario? The default expiration setting will change to none? We probably should document this change. @milosgajdos

**Review by wy65701436 — APPROVED** [review]
lgtm

File: docs/configuration.md
Comment (milosgajdos) [reply docs/configuration.md:1121]: So this entry is used for generating the docs, actually. How we _notify/warn_ users about this change is a whole another question. But I can see this turning a bit tricky. I had a read through related issues [here](https://github.com/distribution/distribution/issues/2367) and [here](https://github.com/distribution/distribution/issues/3247) and it seems that people either:
1. Don't set `STORAGE_CACHE_BLOBDESCRIPTOR` when local file store is used
2. Build a custom image and set expiry `TTL` to a very large number

**Review by waynr — APPROVED** [review]
:see_no_evil:

File: docs/configuration.md
Comment (joaodrp) [reply docs/configuration.md:1121]: I think we should make it default to 7 days, otherwise, it's a breaking change IMO. People using this in production have set up their environments according to the default behavior, changing that without a major release is a risk IMO.

Are there any downsides in making it default to 7 days? Anyone that's not happy with that can override it, and we could use a value of `-1` to disable it.

Comment (joaodrp) [reply docs/configuration.md:1121]: > Are there any downsides in making it default to 7 days?

I see that "disabled by default" can mitigate #2367, but anyone facing that issue could still do it with `ttl: -1`.

Comment (deleteriousEffect) [reply docs/configuration.md:1121]: > Anyone that's not happy with that can override it, and we could use a value of -1 to disable it.

Curious if we could make the type for the TTL a pointer instead, we should be able to tell 0 from not configured with a nil check then and not have to have special values.

Comment (milosgajdos) [reply docs/configuration.md:1121]: @manishtomar WDYT?

Comment (manishtomar) [reply docs/configuration.md:1121]: Since the default behavior is effectively broken (in a way that is difficult to debug) I feel changing the default behavior is the right thing to do. We are still making operators life difficult by silently failing to pull images after 7 days. Currently everyone who encounters the bug has to rebuild the code with high TTL and deploy it. The only thing this PR will do is avoid rebuilding from source. They will _still_ have to face the error, google it and disable TTL to use it. Considering that the end state is always disabling TTL to make the system work why not make it default? FWIW while it seems bad to make backward-incompatible operator changes we've historically [done that in the past](https://github.com/distribution/distribution/issues/2367#issuecomment-454805687) with minor version upgrades. That is not a justification to make another one but seems justifiable in this case.

**Comment by SteveLasker** [comment]
Can we make the default consistent with previous behavior in 2.7 (7-days), and have an option to change the behavior in 2.7 through config?
This way, someone that updates distribution on top of their current release, nothing changes. But, they have the option to change the TTL through config.

Then, in 3.0, we can change the default.

File: docs/configuration.md
Comment (milosgajdos) [reply docs/configuration.md:1121]: Thanks, @manishtomar o/ I think we all agree here re: the default behavior being broken and our having to change it. The "issue" is not so much the change of the current behavior on `main` branch, which no one would object to going forward, but rather that we'd like to propagate this change into the next `2.7.x` patch release and would like to avoid changing the existing behavior, rather than tracking another patch. Would you consider changing this PR to what @SteveLasker suggests in his [comment ](https://github.com/distribution/distribution/pull/3238#issuecomment-827660735)?

**Comment by xavipanda** [comment]
we also experiencing this somehow... it is not very frequent, it takes place maybe a couple times a week in a large fleet.. but when it happens it requires a restart of the pod. Since it seems to have backwards compatibility, what are we needing to move this forward ?

**Comment by prune998** [comment]
There is something that I don't get.
I'm using this Docker Registry as a mirror/cache of DockerHub. Pretty basic setup.
After few months running, I see the disk space used always growing.

So, it seems there's a TTL set to 7 days by default that you want to make configurable.
But, anyways, the proxy functions to delete the manifests and blobs are still not doing anything:

```
func (pms proxyManifestStore) Delete(ctx context.Context, dgst digest.Digest) error {
	return distribution.ErrUnsupported
}
```
Am I missing something here ? What's the point of all this ? 

Thanks

File: docs/configuration.md
Comment (milosgajdos) [reply docs/configuration.md:1121]: Ping @manishtomar

**Comment by liangyuanpeng** [comment]
Really want to know what is missing in the PR before it can be merged.

**Comment by liangyuanpeng** [comment]
The problem is Existed for too long and maintainers seems like is not care about it. 

I'm working on this feature of ttl blobs and I will release it in [liangyuanpeng/distribution](https://github.com/liangyuanpeng/distribution). welcom everyone to work for it together.

**Comment by milosgajdos** [comment]
Heym @liangyuanpeng 
> The problem is Existed for too long and maintainers seems like is not care about it.

This is fundamentally wrong conclusion. Maintainers DO care about this problem (see the discussion we had with the author of this PR), we just don't want to be breaking the existing behaviour for users. We've asked the author of the original PR to make some adjustments.

If you create a PR that addresses the points we've raised in the discussion we'll be more than happy to review it and merge it into mainstream.

**Comment by liangyuanpeng** [comment]
@milosgajdos  Thanks for your reply and happy for my stupid conclusion. I would like to push this PR.

**Comment by wzshiming** [comment]
I want to move forward with this PR, based on which I have made sure of compatibility, please take a look #3880.

**Comment by milosgajdos** [comment]
Closing, addressed in https://github.com/distribution/distribution/pull/3880

**Pull request closed without merging**
